docs/design/constraints.md

```diff
@@ -0,0 +1,177 @@
+- the template which is not allowed in the context will not be shown in `dotnet new list` output
+   - additional option (`--force`) may be considered to show all the templates anyway
+- the template which is not allowed in the context will not be allowed to be instantiated (unless `--force` is specified).
+- the templates that are not allowed to be run will be installed anyway, but in case the template cannot be run, the message will be shown after installation.
+- tab completion: tab completion should not suggest the values from restricted templates. Due to context evaluation for some constraints may take time, time restrictions should be considered(?).
+
+Out of scope:
+- adaptation of `dotnet new search` to list the constraints
+
+## Constraints configuration
+Constraints should be configured in `template.json` in `constraints` section.
+`Constraints` section is JSON object type, containing 1 or more constraint definition key-value pairs. 
+```json
+  "constraints": {
+       "constraintOS": {
+           "type": "os",
+           "args": ["MacOS", "Linux"]      //only allowed on MacOS and Linux
+       },
+       "constraintVS": {
+           "type": "vs-components",
+           "args": {
+                "components": [            //requires Xamarin of version 1.0.0.0 - version 5.0.0.0.
+                    {
+                        "name": "Xamarin",
+                        "minVersion": "1.0.0.0",
+                        "maxVersion": "5.0.0.0"
+                    }
+                ],
+                "whatever": true 
+           }
+       }
+  }
+```
+Constraint definition key is a unique name. The value is an object that has two mandatory keys:
+- `type`: string. Constraint type. All implemented constraints ideally should have unique types. The type will be used to identify constraint component (as alternative to using guids). Case-sensitive.
+- `args`: array or object. Constraint arguments. Arguments depends on constraint implementation. The JSON array or object will be passed to constraint component for evaluation as string. The constraint implementation should be able to parse the content.
+
+Constraint configuration will be read when `template.json` is read and extracted to template engine cache. 
+It will be available in `ITemplateInfo`, therefore for constraint evaluation `template.json` won't be read, and cached value will be used instead. 
+New property `Constraints` will be added to `ITemplateInfo` of type `IReadOnlyList<TemplateConstraintInfo>` containing template constraints information.
+
+`TemplateConstraintInfo` has the following properties:
+- `string Type`: constraint type
+- `string Args`: JSON string containing args read from `template.json`. Args to be parsed when evaluating the constraint by constraint itself.
+
+### Types uniqueness consideration
+At the moment 3rd party constraints won't be supported. 1st party developers should ensure unique naming.
```

Comment:
"won't be supported by dotnet new" would be more accurate right? I could see us supporting 3rd party VS-only constraints, even though it is unlikely. Obviously we would need to solve deconfliction.

Reply:
It is planned to support 3rd party constraints in the future, but not for this release. We still need to clean the acquisition workflow for 3rd party constraints. I would like to postpone any work needed for this to later time.